lib/auth/auth_with_roles.go

```diff
@@ -1129,17 +1129,26 @@ func (a *AuthWithRoles) ChangePasswordWithToken(ctx context.Context, req ChangeP
 	return a.authServer.ChangePasswordWithToken(ctx, req)
 }
 
-// CreateUser inserts a new user entry in a backend.
-func (a *AuthWithRoles) CreateUser(ctx context.Context, user services.User) error {
+// UpsertUserV2 updates an existing user or creates a new user entry in a backend.
+// Captures the user who is modifying this user.
+// Returns a bool that indicates if user existed or not.
+func (a *AuthWithRoles) UpsertUserV2(ctx context.Context, user services.User) (bool, error) {
 	if err := a.action(defaults.Namespace, services.KindUser, services.VerbCreate); err != nil {
-		return trace.Wrap(err)
+		return false, trace.Wrap(err)
+	}
+	if err := a.action(defaults.Namespace, services.KindUser, services.VerbUpdate); err != nil {
+		return false, trace.Wrap(err)
 	}
 
-	user.SetCreatedBy(services.CreatedBy{
-		User: services.UserRef{Name: a.user.GetName()},
-	})
+	if user.GetCreatedBy().IsEmpty() {
+		user.SetCreatedBy(services.CreatedBy{
+			User: services.UserRef{Name: a.user.GetName()},
+		})
+	}
+
+	updatedCtx := WithUpdateBy(ctx, a.user.GetName())
```

Comment:
prefer to re-use the name `ctx`, unless you need both the wrapped and un-wrapped context.
that way it's harder to accidentally use the un-wrapped context in the code that follows.

Reply:
revised